### DISCUSSION TITLE
Support/Documentation Questions

### DISCUSSION THREAD

**Discussion by JRBowman** [VelcroPhysics/VelcroPhysics#73] [Q&A] [answered]
Hello!

I've integrated Velcro Physics into an ECS I've built for MonoGame. Since the beginning I've had some challenges using Dynamic Bodies effectively, primarily around Gravity response and Gravity while the body is applying forces resulting in odd behaviors / movement that just doesn't seem right!

Is there official documentation around the setup of the Physics World object and Forces? I'd like to make sure I'm using the system correctly and get the simulation straightened out!

For further context around where the issues may be stemming from, I have a Pixel-Perfect render pipeline that uses an "internal render resolution" before scaling that pixel-perfect result up to the full screen resolution. I've set the World Simulation to operate at the internal render resolution (since that seemed appropriate) with the Display to Sim Units Ratio being around 28 Pixels-Per-Unit.

Are there any specifics about the Velcro+MonoGame implementation I can read into? I've started reading the source code for Velcro as well as Box2D to get a better idea of what's going on, but any official Velcro Docs would be much appreciated!

Thank you!

**Comment by Genbox** [comment]
Velcro is a combination of Box2D (with some perf improvements) and a lot of helpful tools to build/manipulate polygons. [Box2D has great docs already](https://box2d.org/documentation/), so I don't see the need to replicate them since 95% of the terms/subjects are transferrable.

There are a few best-pratices you should follow:
- Don't stack heavy blocks on light blocks. The solver don't like it too much.
- Run the physics simulation at a ratio of the rendering. Checkout the samples and testbed to see how I've done it there.
- Feel free to use the framework from the samples or testbed to render a debug view. It renders joints, centroids etc. too making it easier to diagnose when something is wrong.
- Don't teleport objects. That is, don't set `body.Position = new Vector2(x, y)`. The solver gets confused and you limit/disable the engines anti-tunneling features.

Could you describe the issues you have? Maybe [record a video](https://gifcap.dev/) of what's happening.

**Reply by JRBowman** [reply]
Thank you for the reply! I figured it was fully transferrable between Box2D and Velcro, I was mostly curious about the specifics of C# + MonoGame Implementation and things to watch out for like you listed for Best Practices!

I saw your method in the demo/samples of rendering the Physics World at a ratio of rendering, I have something similar setup in a `PhysicsSubSystem` class for the ECS, where it's purpose is to accumulate delta time from the Update/Rendering cycle and execute the Physics World.Step() on a Fixed TimeStep of 1/60 (the rendering is not using MonoGame's FixedTimeStep, attempting to enable VRR and unlocked framerates) and then invoke a FixedUpdate(float fixedDeltaTime) for any physics-based component/Entity in the framework 

My setup with the PhysicsSubSystem seems good/reliable and scalable, the only real issues are that objects using Dynamic Bodies in the simulation just don't seem to move like I'd expect - Gravity for instance, it's like no matter what I set the gravity to, the Dynamic Bodies are floating around like molasses haha! If I set the gravity to a really high value, things start falling like expected but then Force-based movement seems to act odd, while moving in the X direction, trying to add any kind of force to jump on the Y axis (currently building a platformer with this ECS) makes the Dynamic Body disproportionately slow down by a lot on X axis and then the body has very floaty movement where it seems to defy gravity

I don't have a video handy but could certainly make one! In the meantime while I was trying to figure out Dynamic Bodies, I ended up basically overriding the DynamicBody behavior and treated it like a KinematicBody with my Kinematic code.

**Reply by JRBowman** [reply, marked as answer]
I got it figured out - There were signs that the problem was scaling, I thought I had taken care of Pixels to Unit Conversions and my test was using the DebugView to validate... well! Like many code tests, the test wasn't working right so I refactored the test until it passed (basically haha) Thinking, "surely the code is right, everything works besides the test"

TLDR; the DebugView didn't seem to align with my world, I thought I had taken care of pixels/coordinates conversions to Units so I 'fixed' the Matrix for the DebugView so it aligned with the world... turns out, I missed some conversion steps for Pixels-Per-Unit! The Physics System and Dynamic Bodies are now working properly with standard gravity physics!

**Reply by JRBowman** [reply]
Also, the only time I set the physics body's transform is when an actual Teleport needs to happen - I try to be safe with it, disabling the Physics Body and using PhysicsBody.SetTransform(...), otherwise everything is pure physics/force movement (Teleporting occurs during world/cell changes.)

I did have 2 more questions for you though, is there an easy way to join/merge Rectangular Fixtures in a TileMap-based system? (I've attempted to use various manual methods tracking shape vertices as well as a few of the built-in polygon tools you provide in Velcro and have had trouble getting anything to work easily)

Is there any built-in method of supporting Liquids? (not liquid simulation or anything, just simplified fluid properties?) - I have implemented my own buoyancy and drag calculations, so not a problem if there isn't anything built-in!

**Reply by Genbox** [reply]
Glad you fixed your problem. The debug view has come in handy more times than I can count. As for your questions:

**Is there an easy way to join/merge Rectangular Fixtures?**
You can attach as many shapes to a body as you want with fixtures. You don't have to merge them to a single set of points (as a polygon), you can just fixture them to a body. However, if you do need to merge several shapes, you can use the ConvexHull tools to "wrap" a polygon around your shapes in order to merge them.

**Is there any built-in method of supporting Liquids?**
There is a BuoyancyController that can create a simple buoyancy effect for bodies. It makes the bodies float in an oscillating fashion to simulate a liquid. Checkout the [BuoyancyTest](https://github.com/Genbox/VelcroPhysics/blob/master/src/VelcroPhysics.MonoGame.Samples.Testbed/Tests/Velcro/BuoyancyTest.cs) in the Testbed.

**Reply by JRBowman** [reply]
Fantastic! Thank you!

The DebugView has been very helpful for me as well, I just accidentally made it lie to me with the view matrix haha

I was just thinking about the merging of rectangles when reading your reply, I guess I can just listen to the PhysicsBody.OnCollision event and get the result I'm thinking of (I hope) I wasn't sure if having many smaller rectangles that are part of a single PhysicsBody would give me grief when trying to determine things like the Fluid/Liquid in my manual code since the body needs to know how much overlap there is for displacement/buoyancy, I was unsure if it was going to reset how much of the body is submerged per Fixture or on the Body as a whole

That's fantastic as well for the Controllers, I started looking into making some custom Controllers but didn't get too far since I was ironing out the rest of the ECS - I'll check out the Testbed code to see implementation details for Controllers and get those added, it sounds like Controllers could add a lot of pluggable functionality I've been doing in manually in other areas of the codebase!

Thanks again!
